app/src/main/java/com/my/boilerplate/StickerBundleProvider.java

```diff
@@ -0,0 +1,359 @@
+                    final long id = db.insert(StickerBundle.Entry.TABLE_NAME,
+                                              null,
+                                              values);
+                    final Uri newUri = getBundleInsertionUriById(id);
+
+                    // Notify the subscribers.
+                    mResolver.notifyChange(newUri, null);
+
+                    return newUri;
+                }
+            }
+        } catch (SQLiteCantOpenDatabaseException ignored) {
+            // DO NOTHING.
+        }
+        return uri;
+    }
+
+    @Override
+    public int delete(@NotNull Uri uri,
+                      String selection,
+                      String[] selectionArgs) {
+        return 0;
+    }
+
+    @Override
+    public String getType(@NotNull Uri uri) {
+        return null;
+    }
+
+    /**
+     * @return {@link #ACTION_QUERY_STICKER_BUNDLE_ALL} |
+     * {@link #ACTION_QUERY_STICKER_BUNDLE}
+     */
+    @URI_ACTION
+    public static int getUriAction(Uri uri) {
+        @URI_ACTION int action = sUriMatcher.match(uri);
+        return action;
+    }
+
+    public static Uri getBundlePrefixUri() {
+        return Uri.parse(SCHEME + AUTHORITY + PATH_STICKER);
+    }
+
+    public static Uri getBundleListUri() {
+        return Uri.parse(SCHEME + AUTHORITY + PATH_QUERY_STICKER_ALL);
+    }
+
+    public static Uri getBundleInsertionUri() {
+        return Uri.parse(SCHEME + AUTHORITY + PATH_INSERT_STICKER);
+    }
+
+    public static Uri getBundleInsertionUriById(long id) {
+        return Uri.parse(SCHEME + AUTHORITY + PATH_INSERT_STICKER + "/" + id);
+    }
+
+    public static ContentValues bundleToContentValues(StickerBundle bundle) {
+        ContentValues values = new ContentValues();
+
+        // TODO: The uniform database for different version?
+        values.put(StickerBundle.Entry.COL_NAME, bundle.name);
+        values.put(StickerBundle.Entry.COL_PRICE, bundle.price);
+        values.put(StickerBundle.Entry.COL_COUNT, bundle.count);
+        values.put(StickerBundle.Entry.COL_COVER, bundle.coverUrl);
+        values.put(StickerBundle.Entry.COL_TITLE, bundle.title);
+        values.put(StickerBundle.Entry.COL_DESC, bundle.description);
+        values.put(StickerBundle.Entry.COL_INSTALL_URL, bundle.installSourceUrl);
+        values.put(StickerBundle.Entry.COL_STICKER_THUMBS,
+                   StringUtil.join(StringUtil.COMMA, bundle.stickerThumbnails));
+        // TODO: Need install_requirement and promotion_info
+
+        return values;
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Protected / Private Methods ////////////////////////////////////////////
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Clazz //////////////////////////////////////////////////////////////////
+
+    private static class StickerBundleDbHelper extends SQLiteOpenHelper {
```

Comment:
@bearprada The reason I make the "helper" class as an internal class in the "provider" class is that the "helper" is actually used only by the "provider". They have such coupled relationship. And for the external callers, they should operate the "provider" directly with comprehensive URI APIs.